### PR TITLE
Delayed externalize metric

### LINKED ISSUE #2451: consider improving metrics that report "delayed" validators
Right now "delayed" is reported in a very opaque way: any message received post externalize is considered "delayed".

The problem is that we don't actually get a good idea of the variance of those delays: some validators could be milliseconds behind, where as others could be seconds behind.

### PR DESCRIPTION
**Pull request by MonsieurNicolas**
# Description

Resolves #2451

This PR adds two new metrics to help track
* the "perceived lag" between network closing (modeled as delay between "first externalize event" and actual externalize)
* validators that are "behind" (modeled as delay between the node externalizing and receiving externalize messages from other nodes)

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/herder/HerderSCPDriver.h [hunk 1]
```diff
@@ -88,6 +88,8 @@ class HerderSCPDriver : public SCPDriver
 
     void recordSCPExecutionMetrics(uint64_t slotIndex);
     void recordSCPEvent(uint64_t slotIndex, bool isNomination);
+    void recordSCPExternalizeEvent(uint64_t slotIndex, NodeID const& id,
+                                   bool forceupdateself);
```
Comment (marta-lokhova) [inline src/herder/HerderSCPDriver.h]: nit: we typically use camel case

File: src/herder/HerderSCPDriver.cpp [hunk 2]
```diff
@@ -933,6 +945,48 @@ HerderSCPDriver::recordSCPEvent(uint64_t slotIndex, bool isNomination)
...
+    {
+        timing.mFirstExternalize = make_optional<VirtualClock::time_point>(now);
+    }
+
+    bool self = (id == mSCP.getLocalNodeID());
+    if (self)
+    {
+        if (!timing.mSelfExternalize)
+        {
+            if (timing.mFirstExternalize)
```
Comment (marta-lokhova) [inline src/herder/HerderSCPDriver.cpp]: `timing.mFirstExternalize` is either present already or set on line 959, so I don't think you need this check.

File: docs/metrics.md [hunk 3]
```diff
@@ -116,8 +116,10 @@ scp.pending.ready                        | counter   | number of envelopes ready
 scp.sync.lost                            | meter     | validator lost sync
 scp.timeout.nominate                     | meter     | timeouts in nomination
 scp.timeout.prepare                      | meter     | timeouts in ballot protocol
-scp.timing.externalized                  | timer     | time spent in ballot protocol
 scp.timing.nominated                     | timer     | time spent in nomination
+scp.timing.externalized                  | timer     | time spent in ballot protocol
+scp.timing.externalize-lag               | timer     | delay between first externalize message received and node externalizing
+scp.timing.externalize-delay             | timer     | delay between node externalizing and other externalize messages
```
Comment (marta-lokhova) [inline docs/metrics.md]: Can we slightly tweak this description to make it clear that "other externalize messages" are the externalize messages received after the node itself externalizes?

File: src/herder/HerderSCPDriver.cpp (hunk 2 above)
Comment (MonsieurNicolas) [reply src/herder/HerderSCPDriver.cpp]: yeah I wanted to use `first` here (so bug), but I think it's better to always capture. so I removed this condition

File: docs/metrics.md (hunk 3 above)
Comment (MonsieurNicolas) [reply docs/metrics.md]: done

Comment (marta-lokhova) [reply docs/metrics.md]: Looks like the description is still the same?

Comment (MonsieurNicolas) [reply docs/metrics.md]: I thought you were asking about "node externalizing" that I clarified as "local node externalizing", but yeah I replaced "other" with "later"

**Comment by marta-lokhova** [comment]
r+ 8b0e5d9de72168e04c9f5d51b9621c02594c00db

**Pull request merged**